deepchem/rl/mcts.py

```diff
@@ -0,0 +1,411 @@
+    root = TreeSearchNode(0.0)
+    for step in range(steps):
+      prob, reward = self._do_tree_search(root, temperature, adapt_puct)
+      state = self._env.state
+      if not self._state_is_list:
+        state = [state]
+      buffer.append((state, prob, reward))
+      action = np.random.choice(np.arange(self._env.n_actions), p=prob)
+      self._env.step(action)
+      if self._env.terminated:
+        self._env.reset()
+        root = TreeSearchNode(0.0)
+      else:
+#        root = root.children[action]
+        root = TreeSearchNode(0.0)
+      if time.time() > self._checkpoint_time:
+        self._checkpoint_index += 1
+        saver.save(
+            self._graph.session, self._graph.save_file, global_step=self._checkpoint_index)
+        self._checkpoint_time = time.time()
+    return buffer
+
+  def _optimize_policy(self, buffer, epochs):
+    """Optimize the policy based on the replay buffer from the current iteration."""
+    batch_size = self._graph.batch_size
+    n_batches = len(buffer) // batch_size
+    for epoch in range(epochs):
+      np.random.shuffle(buffer)
+
+      def generate_batches():
+        for batch in range(n_batches):
+          indices = list(range(batch*batch_size, (batch+1)*batch_size))
+          feed_dict = {}
+          for i, f in enumerate(self._features):
+            feed_dict[f] = np.stack(buffer[j][0][i] for j in indices)
+          feed_dict[self._search_prob] = np.stack(buffer[j][1] for j in indices)
+          feed_dict[self._search_value] = np.array([buffer[j][2] for j in indices])
+          yield feed_dict
+
+      loss = self._graph.fit_generator(generate_batches(), checkpoint_interval=0)
+
+  def _do_tree_search(self, root, temperature, adapt_puct):
+    """Perform the tree search for a state."""
+    # Build the tree.
+
+    for i in range(self.n_search_episodes):
+      env = copy.deepcopy(self._env)
+      trace = []
+      self._create_trace(env, root, trace)
+      self._record_trace_rewards(trace)
+
+    # Compute the final probabilities and expected reward.
+
+    prob = np.array([c.count**(1.0/temperature) for c in root.children])
+    prob /= np.sum(prob)
+    reward = np.sum(p*c.mean_reward for p, c in zip(prob, root.children))
+    if adapt_puct:
+      scale = np.sum([p*np.abs(c.mean_reward) for p, c in zip(prob, root.children)])
+      self._puct_scale = 0.99 * self._puct_scale + 0.01 * scale
+    return prob, reward
+
+  def _create_trace(self, env, node, trace):
+    """Create one trace as part of the tree search."""
+    trace.append(node)
+    node.count += 1
+    if env.terminated:
+      # Mark this node as terminal
+      node.children = None
+      node.value = 0.0
+      return
+    if node.children is not None and len(node.children) == 0:
+      # Expand this node.
+      prob_pred, value = self.predict(env.state)
+      node.value = float(value)
+      node.children = [TreeSearchNode(p) for p in prob_pred[0]]
+    if len(trace) == self.max_search_depth:
```

Comment:
To be clear, does this generate self.max_search_depth + 1 layers in search tree? I was wondering why you didn't check this first before expanding.

Reply:
If we've reached the maximum search depth, we need to set `node.value` as done just above.  Since we're not going to search any further, we fill in our best estimate of the rewards we *would* have gotten by continuing.